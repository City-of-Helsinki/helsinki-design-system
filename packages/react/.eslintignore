**/*.d.ts
webpack.config.js
.templates
lib/
**/*.stories.tsx
