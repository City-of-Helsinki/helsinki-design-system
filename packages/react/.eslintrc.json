{
  "parser": "@typescript-eslint/parser",
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:react/recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:import/errors",
    "plugin:import/warnings",
    "plugin:import/typescript",
    "plugin:jest/recommended",
    "airbnb",
    "prettier/@typescript-eslint",
    "plugin:prettier/recommended" // this should always be the last item on this array
  ],
  "plugins": ["react", "@typescript-eslint", "prettier", "jest", "import"],
  "parserOptions": {
    "ecmaVersion": 2018,
    "ecmaFeatures": {
      "jsx": true
    }
  },
  "ignorePatterns": ["*.module.css"],
  "rules": {
    "no-use-before-define": "off",
    "@typescript-eslint/no-use-before-define": ["error"],
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": ["**/*.test.ts", "**/*.test.tsx", "**/test*.ts*", "**/*.stories.tsx"]
      }
    ],
    "import/order": [
      "error",
      {
        "groups": ["builtin", "external", "internal", ["parent", "sibling", "index"]],
        "newlines-between": "always"
      }
    ],
    "react/jsx-filename-extension": "off",
    "react/jsx-one-expression-per-line": "off",
    "react/jsx-props-no-spreading": "off",
    "react/jsx-curly-newline": "off",
    "@typescript-eslint/explicit-function-return-type": "off",
    "react/prop-types": "off",
    "import/extensions": "off",
    "import/prefer-default-export": "off",
    "jsx-a11y/label-has-associated-control": [
      2,
      {
        "components": ["Label"],
        "required": {
          "some": ["nesting", "id"]
        },
        "allowChildren": false
      }
    ],
    "@typescript-eslint/explicit-member-accessibility": "off",
    "react/state-in-constructor": [2, "never"],
    "no-unused-expressions": [2, {
      "allowTernary": true
    }],
    "no-unused-vars": "off",
    "@typescript-eslint/no-unused-vars": ["error"]
  },
  "settings": {
    "react": {
      "version": "detect"
    },
    "import/resolver": {
      "node": {
        "extensions": [".js", ".jsx", ".ts", ".tsx"],
        "paths": ["src"]
      }
    }
  },
  "env": {
    "browser": true,
    "jest": true
  },
  "globals": {
    "JSX": true
  }
}
