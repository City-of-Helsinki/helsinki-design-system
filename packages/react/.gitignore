node_modules/
storybook-static/
lib/
_bundles/
*.css.d.ts
.loki/current
.loki/difference

*storybook.log
