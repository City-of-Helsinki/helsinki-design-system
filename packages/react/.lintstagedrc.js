module.exports = {
  'src/**/*.{ts,tsx}': 'eslint',
}
