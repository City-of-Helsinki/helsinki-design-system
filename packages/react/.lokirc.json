{
  "chromeRetries": 2,
  "configurations": {
    "chrome.laptop": {
      "target": "chrome.docker",
      "width": 1366,
      "height": 768,
      "mobile": false
    },
    "chrome.iphone7": {
      "target": "chrome.docker",
      "preset": "iPhone 7"
    }
  },
  "skipStories": "Icons/.*|Button/*|Tag/*|Hero/*|Koros/*|Checkbox/*",
  "pixelmatch": {
    "threshold": 0.2,
    "alpha": 1,
    "diffColor": [255, 0, 255]
  }
}
