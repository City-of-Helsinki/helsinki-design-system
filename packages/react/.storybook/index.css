@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/565d73a693abe0776c801607ac28f0bf.woff') format('woff');
  font-display: swap;
  font-weight: 400;
  font-style: normal;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/5bb29e3b7b1d3ef30121229bbe67c3e1.woff') format('woff');
  font-display: swap;
  font-weight: 400;
  font-style: italic;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/7c46f288e8133b87e6b12b45dac71865.woff') format('woff');
  font-display: swap;
  font-weight: 500;
  font-style: normal;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/e62dc97e83a385e4d8cdc939cf1e4213.woff') format('woff');
  font-display: swap;
  font-weight: 500;
  font-style: italic;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/533af26cf28d7660f24c2884d3c27eac.woff') format('woff');
  font-display: swap;
  font-weight: 700;
  font-style: normal;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/20d494430c87e15e194932b729d48270.woff') format('woff');
  font-display: swap;
  font-weight: 700;
  font-style: italic;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/a50a1bd245ce63abcc0d1da80ff790d2.woff') format('woff');
  font-display: swap;
  font-weight: 900;
  font-style: normal;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/62a1781d8b396fbb025b0552cf6304d2.woff') format('woff');
  font-display: swap;
  font-weight: 900;
  font-style: italic;
  text-rendering: optimizeLegibility;
}

body {
  font-family: var(--font-default);
  font-size: var(--fontsize-body-m);
  color: var(--color-black-90);
  line-height: var(--lineheight-l);
}

.password-input--external-show-password-button,
.date-input--external-clear-value-button {
  display: grid;
  grid-template-columns: 400px 200px;
  gap: 20px;
  align-items: center;
}

@media only screen and (max-width: 768px) {
  .password-input--external-show-password-button,
  .date-input--external-clear-value-button {
    grid-template-columns: 300px;
  }
}

@media only screen and (max-width: 575.98px) {
  .table-custom-action {
    width: 100%;
  }
}

/* STEPPER */

.stepper-small {
  max-width: 300px;
}

@media only screen and (max-width: 575.98px) {
  .stepper-small {
    max-width: 100%;
  }
}

.stepper-margin {
  margin-left: -35px;
}

.stepper-heading {
  margin-left: 10px;
}

.stepper-form-validation {
  margin-left: -18px;
}

.stepper-card > div:first-child {
  margin-bottom: var(--spacing-m) !important;
}

#focused-element-after-cookie-consent-closed:focus {
  outline: 2px black solid;
}
