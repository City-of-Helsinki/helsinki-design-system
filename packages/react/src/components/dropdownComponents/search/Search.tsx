import { uniqueId } from 'lodash';
import React, { useMemo, createRef, useEffect, forwardRef, useCallback } from 'react';

import { SelectProps, SelectMetaData, SelectData } from '../select/types';
import { AcceptedNativeDivProps, Option } from '../modularOptionList/types';
import { Container } from '../select/components/Container';
import { Label } from '../select/components/Label';
import { changeHandler } from '../select/dataUpdater';
import { getElementIds } from '../select/utils';
import { DataProvider, DataProviderProps } from '../../dataProvider/DataProvider';
import { SelectionsAndListsContainer } from '../select/components/SelectionsAndListsContainer';
import { ModularOptionList } from '../modularOptionList';
import { ListAndInputContainer } from '../select/components/ListAndInputContainer';
import { TagList } from '../select/components/tagList/TagList';
import { ErrorNotification } from '../select/components/Error';
import { AssistiveText } from '../select/components/AssistiveText';
import { createTextProvider } from '../select/texts';
import { eventIds } from '../select/events';
import { ScreenReaderNotifications } from '../select/components/ScreenReaderNotifications';
import { getSelectedOptions, convertPropsToGroups, mutateGroupLabelSelections } from '../modularOptionList/utils';
import { SearchInput } from '../select/components/searchInput/SearchInput';

export const Search = forwardRef<HTMLButtonElement, Omit<SelectProps & AcceptedNativeDivProps, 'ref'>>(
  (
    {
      options,
      open,
      groups,
      icon,
      required,
      onChange,
      children,
      id,
      onFocus,
      onBlur,
      onClose,
      disabled,
      texts,
      invalid,
      multiSelect,
      noTags,
      visibleOptions,
      virtualize,
      filter,
      onSearch,
      value,
      theme,
      clearable,
      tooltip,
      ...divElementProps
    },
    ref,
  ) => {
    const initialData = useMemo<SelectData>(() => {
      const data = {
        groups: convertPropsToGroups({ groups, options, value, children }),
        open: !!open,
        required: !!required,
        invalid: !!invalid,
        disabled: !!disabled,
        multiSelect: !!multiSelect,
        noTags: !!noTags,
        visibleOptions: visibleOptions || 5.5,
        virtualize: !!virtualize,
        onChange,
        onFocus,
        onBlur,
        onClose,
        filterFunction: filter,
        onSearch,
        clearable: !!clearable,
        initialOpenValue: open,
      };
      if (data.multiSelect) {
        mutateGroupLabelSelections(data.groups);
      }
      return data;
    }, [
      options,
      open,
      groups,
      onChange,
      disabled,
      invalid,
      required,
      noTags,
      virtualize,
      visibleOptions,
      onSearch,
      onFocus,
      onBlur,
      onClose,
      value,
      children,
      clearable,
    ]);

    const metaData = useMemo((): SelectMetaData => {
      const containerId = `${id || uniqueId('hds-select-')}`;
      const optionIds = new Map<string, string>();
      let optionIdCounter = 0;
      const getListInputType = () => {
        if (!initialData.onSearch && !initialData.filterFunction) {
          return undefined;
        }
        return initialData.onSearch ? eventIds.search : eventIds.filter;
      };
      return {
        lastToggleCommand: 0,
        lastClickedOption: undefined,
        showAllTags: false,
        icon,
        activeDescendant: undefined,
        focusTarget: undefined,
        refs: {
          button: typeof ref === 'function' ? createRef<HTMLButtonElement>() : ref || createRef<HTMLButtonElement>(),
          listContainer: createRef<HTMLDivElement>(),
          list: createRef<HTMLUListElement>(),
          selectionsAndListsContainer: createRef<HTMLDivElement>(),
          tagList: createRef<HTMLDivElement>(),
          showAllButton: createRef<HTMLButtonElement>(),
          searchOrFilterInput: createRef<HTMLInputElement>(),
          container: createRef<HTMLDivElement>(),
        },
        selectedOptions: getSelectedOptions(initialData.groups),
        elementIds: getElementIds(containerId),
        textProvider: createTextProvider(texts),
        getOptionId: (option: Option) => {
          const identifier = option.isGroupLabel ? `hds-group-${option.label}` : option.value;
          const current = optionIds.get(identifier);
          if (!current) {
            const optionId = `${containerId}-option-${optionIdCounter}`;
            optionIdCounter += 1;
            optionIds.set(identifier, optionId);
            return optionId;
          }
          return current;
        },
        listInputType: getListInputType(),
        hasListInput: !!getListInputType(),
        filter: '',
        search: '',
        isSearching: false,
        hasSearchError: false,
        cancelCurrentSearch: undefined,
        screenReaderNotifications: [],
        tooltip,
      };
    }, [id, initialData.groups, initialData.filterFunction, initialData.onSearch, texts, ref]);

    useEffect(() => {
      return () => {
        if (metaData.cancelCurrentSearch) {
          metaData.cancelCurrentSearch();
        }
      };
    }, []);

    const onReset: DataProviderProps<SelectData, SelectMetaData>['onReset'] = useCallback(
      ({ previousData, currentData, currentMetaData }) => {
        if (currentData) {
          if (previousData) {
            // if the "open"-prop is explicitly set and has changed, it must be deliberate.
            if (
              typeof currentData.initialOpenValue !== 'undefined' &&
              previousData.initialOpenValue !== currentData.initialOpenValue
            ) {
              return { ...currentData, open: currentData.initialOpenValue };
            }
            // if the list was open prior to the re-render, it should still be
            if (previousData.open) {
              return { ...currentData, open: true };
            }
          }
          return currentData;
        }
        return currentMetaData;
      },
      [],
    );

    return (
      <DataProvider<SelectData, SelectMetaData>
        initialData={initialData}
        metaData={metaData}
        onChange={changeHandler}
        onReset={onReset}
      >
        <Container {...divElementProps} theme={theme}>
          <Label />
          <SelectionsAndListsContainer>
            <SearchInput id={`${id}-searchInput`} />
            <ListAndInputContainer>
              <ModularOptionList {...divElementProps} theme={theme} />
            </ListAndInputContainer>
          </SelectionsAndListsContainer>
          <ErrorNotification />
          <AssistiveText />
          <TagList />
          <ScreenReaderNotifications />
        </Container>
      </DataProvider>
    );
  },
);
