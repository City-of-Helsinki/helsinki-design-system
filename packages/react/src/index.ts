export * from './icons';
export * from './components';
export * from './ssr';
