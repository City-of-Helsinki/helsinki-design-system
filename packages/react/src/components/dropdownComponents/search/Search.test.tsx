import React from 'react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';

import { Search } from './Search';

describe('<Search /> spec', () => {
  it('renders the component', () => {
    const { asFragment } = render(<Search />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should not have basic accessibility issues', async () => {
    const { container } = render(<Search />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
