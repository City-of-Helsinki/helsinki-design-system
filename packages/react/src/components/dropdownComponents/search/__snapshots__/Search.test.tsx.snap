// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> spec renders the component 1`] = `
<DocumentFragment>
  <div
    class="wrapper root"
    id="hds-select-1"
    tabindex="-1"
  >
    <div
      class="selectAndListContainer"
    >
      <div
        class="root searchInput"
      >
        <div
          class="inputWrapper"
        >
          <input
            aria-label="Search"
            class="input hasClearButton"
            id="test"
            placeholder="Anna teksti hakua varten"
            type="text"
            value=""
          />
          <div
            class="buttonWrapper"
          >
            <button
              aria-label="TyhjennÃ¤ haku"
              class="button clearButton"
              type="button"
            >
              <svg
                aria-hidden="true"
                aria-label="cross-circle"
                class="icon small"
                role="img"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M12 2C17.5228 2 22 6.47715 22 12C22 17.5228 17.5228 22 12 22C6.47715 22 2 17.5228 2 12C2 6.47715 6.47715 2 12 2ZM12 4C7.58172 4 4 7.58172 4 12C4 16.4183 7.58172 20 12 20C16.4183 20 20 16.4183 20 12C20 7.58172 16.4183 4 12 4ZM15 7.5L16.5 9L13.5 12L16.5 15L15 16.5L12 13.5L9 16.5L7.5 15L10.5 12L7.5 9L9 7.5L12 10.5L15 7.5Z"
                  fill="currentColor"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        aria-hidden="true"
        class="listAndInputContainer"
        id="hds-select-1-sl-container"
      />
    </div>
    <div
      aria-live="polite"
      class="screenReaderNotifications"
      data-testid="hds-select-screen-reader-notifications"
    />
  </div>
</DocumentFragment>
`;
